aslam_cv_pipeline/src/visual-npipeline.cc

```diff
@@ -76,37 +79,45 @@ bool VisualNPipeline::processImageNonBlockingDroppingOldestNFrameIfFull(
     size_t max_output_queue_size) {
   CHECK_GE(max_output_queue_size, 1u);
 
-  bool result = false;
+  if (shutdown_) {
+    return false;
+  }
   std::unique_lock<std::mutex> lock(mutex_);
   if (completed_.size() >= max_output_queue_size) {
     completed_.erase(completed_.begin());
-    result = true;
+    condition_not_full_.notify_all();
   }
-  processImage(camera_index, image, timestamp);
-  return result;
+  processImageImpl(camera_index, image, timestamp);
+  return true;
 }
 
 bool VisualNPipeline::getNextBlocking(std::shared_ptr<VisualNFrame>* nframe) {
   CHECK_NOTNULL(nframe);
-  std::unique_lock<std::mutex> lock(mutex_);
-  if (shutdown_) {
-    return false;
-  }
-  if (completed_.empty()) {
-    condition_not_empty_.wait(lock);
+
+  while (!shutdown_) {
+    std::unique_lock<std::mutex> lock(mutex_);
     if (completed_.empty()) {
-      // Queue shutdown in the meantime.
-      return false;
+      condition_not_empty_.wait(lock);
+      if (shutdown_) {
+        return false;
+      }
+      if (completed_.empty()) {
+        continue;
+      }
     }
+    // Get the oldest frame.
+    *nframe = getNextImpl();
+    CHECK(*nframe);
+    return true;
   }
-
-  // Get the oldest frame.
-  *nframe = getNextImpl();
-  CHECK(*nframe);
-  return true;
+  return false;
 }
 
 void VisualNPipeline::processImage(size_t camera_index, const cv::Mat& image, int64_t timestamp) {
+  if (shutdown_) {
+    return;
+  }
+  std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
is the lock necessary here?

Reply:
Yes i think so. what if you call this function directly without having a limit on the queue size?

Reply:
no actually not... good catch

Reply:
@schneith Hmm I don't see how this is relevant for the queue (map) size limit. It will just increase the size once the frame is processed.